### PR TITLE
Fix join filter rewrites with nested queries

### PR DESCRIPTION
**Pull request by jon-wei**
Fixes https://github.com/apache/druid/issues/9792 by moving the join filter pre-analysis into the `makeCursors` method of `HashJoinSegmentStorageAdapter`. This is done by introducing a new `JoinFilterPreAnalysisGroup` class, which holds a concurrent hash map of Filter -> JoinFilterPreAnalysis, used to avoid redundant computation of the JoinFilterPreAnalysis

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by jon-wei** [comment]
Marking WIP, need to adjust javadocs, will also see test coverage results

**Comment by jon-wei** [comment]
Added javadocs and fixed some inspections, removing WIP

File: processing/src/main/java/org/apache/druid/query/Query.java [hunk 1]
```diff
@@ -170,9 +169,4 @@ default String getSqlQueryId()
   {
     return withOverriddenContext(ImmutableMap.of(QueryContexts.LANE_KEY, lane));
   }
-
-  default VirtualColumns getVirtualColumns()
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/Query.java:174]: I'm assuming this change has nothing to do with the bug fix correct?

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java [hunk 2]
```diff
@@ -209,13 +209,36 @@ public Metadata getMetadata()
       @Nullable final QueryMetrics<?> queryMetrics
   )
   {
-    if (!Objects.equals(joinFilterPreAnalysis.getOriginalFilter(), filter)) {
-      throw new ISE(
-          "Filter provided to cursor [%s] does not match join pre-analysis filter [%s]",
+    JoinFilterPreAnalysis jfpa;
+    if (filter == null) {
+      jfpa = JoinFilterAnalyzer.computeJoinFilterPreAnalysis(
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java:214]: Does this mean we'll re-compute the pre-analysis every time if the filter is null?

I think this code path is short-circuited right now, but it could change in the future and it would be hard to remember that the computation is not cached.

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java [hunk 3]
```diff
@@ -209,13 +209,36 @@ public Metadata getMetadata()
...
+                joinFilterPreAnalysisGroup,
+                JoinableClauses.fromList(clauses),
+                virtualColumns,
+                theFilter,
+                joinFilterPreAnalysisGroup.isEnableFilterPushDown(),
+                joinFilterPreAnalysisGroup.isEnableFilterRewrite(),
+                joinFilterPreAnalysisGroup.isEnableRewriteValueColumnFilters(),
+                joinFilterPreAnalysisGroup.getFilterRewriteMaxSize()
+            );
+          }
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java:238]: I think a better abstraction is to hide all of this logic in the JoinFilterPreAnalysisGroup. We shouldn't expose `getAnalyses()` to the other classes.

Instead, consider exposing a function like `getPreAnalysisForFilter(Filter f)` - that provides the `jfpa` Users of the API don't need to worry about how it works as long as we guarantee that this function is thread safe.

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterPreAnalysis.java [hunk 4]
```diff
@@ -51,8 +52,7 @@
   private final List<Filter> normalizedBaseTableClauses;
   private final List<Filter> normalizedJoinTableClauses;
   private final JoinFilterCorrelations correlations;
-  private final boolean enableFilterPushDown;
-  private final boolean enableFilterRewrite;
+  private final JoinFilterPreAnalysisGroup myGroup;
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterPreAnalysis.java:55]: Does this introduce a memory leak / a circular dependency of some kind?

The group holds a reference to the pre-analysis object via the `analyses` concurrent map and the pre-analyses object holds a reference back to the group object via `myGroup`

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java (hunk 3 above)
Comment (suneet-s) [reply processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java:238]: Also, since the pre-analysis computation is expensive, the way this is written, I think it's still possible for all the threads on the historical to attempt to compute the preAnalysis, so we could see a spike in CPU usage.

File: processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * A concurrent hash map is used since the per-level Filter processing occurs across multiple threads.
+ */
+public class JoinFilterPreAnalysisGroup
+{
+  private final boolean enableFilterPushDown;
+  private final boolean enableFilterRewrite;
+  private final boolean enableRewriteValueColumnFilters;
+  private final long filterRewriteMaxSize;
+  private final ConcurrentHashMap<Filter, JoinFilterPreAnalysis> analyses;
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java:42]: Have you looked at the implementations of hashCode and equals for all filters? Are they efficient? What happens if the filter is something like an IN filter with a very large list of values? The hashCode check could be slower than deciding that the filter can not be pushed down.

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java (hunk 3 above)
Comment (suneet-s) [reply processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java:238]: > Also, since the pre-analysis computation is expensive, the way this is written, I think it's still possible for all the threads on the historical to attempt to compute the preAnalysis, so we could see a spike in CPU usage.

Sorry I misread the javadocs. ConcurrentHashMap blocks on an operation on the same key when a computation is on-going, so this shouldn't be an issue.

From the javadocs
> The entire method invocation is performed atomically, so the function is
> applied at most once per key.  Some attempted update operations
> on this map by other threads may be blocked while computation
> is in progress, so the computation should be short and simple,
> and must not attempt to update any other mappings of this map.

File: processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java (hunk 5 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java]: The main one I would be concerned with is InFilter, and to a lesser extent AND/OR.

I added lazy computation/caching for InFilter/AndFilter/OrFilter hashcodes, as long as the hashing is done once per query, and not once per segment, I wouldn't expect hashing overhead to outweigh the benefits of the rewrite (esp since such large filters would be expensive to apply per-row on the RHS)

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterPreAnalysis.java (hunk 4 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterPreAnalysis.java]: I think it's fine, but I restructured this to store all the config parameters in a new `JoinFilterRewriteConfig` class, so there's no more reference to the group in the pre-analysis objects.

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java (hunk 2 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java]: I adjusted the analysis group so the key is now (Filter, JoinableClauses, VirtualColumns) instead of filter, I think that's a more correct key, and the null filters are cached now.

File: processing/src/main/java/org/apache/druid/query/Query.java (hunk 1 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/Query.java]: It was something failing inspections since all the implementations of Query implement that method, it's only needed for the old rewrite mode (so I added this back in)

**Comment by jon-wei** [comment]
> I haven't looked through any of the tests yet. Since there's a lot of tradeoffs to be made, I think we should have a short-circuit in the JoinFilterPreAnalysisGroup that falls back to the old behavior which is disabled by default.

I added a query context option and a separate set of methods for the old rewrite mode to JoinFilterPreAnalysisGroup, described as something available temporarily until the new mode is more battle-tested.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQuery.java [hunk 6]
```diff
@@ -249,7 +249,6 @@ private GroupByQuery(
     return subtotalsSpec;
   }
 
-  @Override
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByQuery.java:252]: Why this change, `Query.java` provides a default implementation, so doesn't it `@Override`?

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 7]
```diff
@@ -264,6 +266,26 @@ public String toString()
...
+   * The join filter rewrites originally performed the pre-analysis phase prior to any per-segment processing,
+   * analyzing only the filter in the top-level of the query.
+   *
+   * This did not work for nested queries (see https://github.com/apache/druid/pull/9978), so the rewrite pre-analysis
+   * was moved into the cursor creation of the {@link org.apache.druid.segment.join.HashJoinSegmentStorageAdapter}.
+   * This design requires synchronization across multiple segment processing threads; the old rewrite mode
+   * is kept temporarily available in case issues arise with the new mode, and the user does not run queries with the
+   * affected nested shape.
+   */
+  public static <T> boolean getUseJoinFilterRewriteOldRewriteMode(Query<T> query)
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:281]: is it possible to detect this and use the different modes automatically since it sounds like the old mode is perhaps better if there are no subqueries involved?

File: processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java [hunk 8]
```diff
@@ -149,28 +153,36 @@ public FilterTuning getFilterTuning()
...
           .build();
     }
     return cacheKey;
   }
 
+  private static HashCode computeValuesHashCode(Set<String> values)
+  {
+    final List<String> sortedValues = new ArrayList<>(values);
+    sortedValues.sort(Comparator.nullsFirst(Ordering.natural()));
+    final Hasher hasher = Hashing.sha256().newHasher();
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/filter/InDimFilter.java]: I know this isn't new, but might be worth investigating if there are faster hashes that are unique enough for this (not necessary in this PR, just thinking out loud)

File: processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      List<JoinableClause> clauses,
+      VirtualColumns virtualColumns
+  )
+  {
+    // Some filters have potentially expensive hash codes that are lazily computed and cached.
+    // We call hashCode() here in a synchronized block before we attempt to use the Filter in the analyses map,
+    // to ensure that the hashCode is only computed once per Filter since the Filter interface is not thread-safe.
+    synchronized (analyses) {
+      if (filter != null) {
+        filter.hashCode();
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java:85]: Oh, is this why we don't need thread-safety on filter hashcode methods i guess? This seems kind of a funny way to prime them with the cached values, I think maybe the supplier.memoize pattern would be a little cleaner and make this not necessary?

File: processing/src/test/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapterTest.java [hunk 10]
```diff
@@ -302,21 +298,13 @@ public void test_makeCursors_factToCountryLeft()
...
-    JoinFilterPreAnalysis preAnalysis = JoinFilterAnalyzer.computeJoinFilterPreAnalysis(
-        JoinableClauses.fromList(joinableClauses),
-        VirtualColumns.EMPTY,
-        null,
-        true,
-        true,
-        true,
-        QueryContexts.DEFAULT_ENABLE_JOIN_FILTER_REWRITE_MAX_SIZE
-    );
+    JoinFilterPreAnalysisGroup joinFilterPreAnalysisGroup = makeDefaultConfigPreAnalysisGroup();
```
Comment (clintropolis) [inline processing/src/test/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapterTest.java:301]: nice

File: processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java (hunk 3 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java]: I've added a `computeJoinFilterPreAnalysisIfAbsent` and `getAnalysis` method to `JoinFilterPreAnalysisGroup`

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 11]
```diff
@@ -11486,79 +11485,75 @@ public void testTimeExtractWithTooFewArguments() throws Exception
   @Parameters(source = QueryContextForJoinProvider.class)
   public void testNestedGroupByOnInlineDataSourceWithFilterIsNotSupported(Map<String, Object> queryContext) throws Exception
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: Ah, fixed test name

File: processing/src/main/java/org/apache/druid/query/filter/AndDimFilter.java [hunk 12]
```diff
@@ -142,7 +147,10 @@ public boolean equals(Object o)
   @Override
   public int hashCode()
   {
-    return fields != null ? fields.hashCode() : 0;
+    if (fieldsHashCode == null) {
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/filter/AndDimFilter.java:150]: I changed this and elsewhere to use Supplier.memoize

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQuery.java (hunk 6 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByQuery.java:252]: This was a change from removing the "old rewrite mode" initially, but now that it's added back in, so I put the overrides back

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java (hunk 7 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/QueryContexts.java:281]: Hmm, I think that could be worth looking into later on

File: processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java (hunk 9 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java]: Removed this after changing to supplier.memoize

File: processing/src/main/java/org/apache/druid/segment/filter/AndFilter.java [hunk 13]
```diff
@@ -257,12 +264,12 @@ public boolean equals(Object o)
       return false;
     }
     AndFilter andFilter = (AndFilter) o;
-    return Objects.equals(getFilters(), andFilter.getFilters());
+    return Objects.equals(hashCode(), andFilter.hashCode());
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/filter/AndFilter.java:267]: equals should not compare hashCodes. It's possible that different and filters hash to the same hashCode.

Similar comments in the other filters

**Comment by jon-wei** [comment]
I'm going to close this one and open a new PR with a different approach, this one has conflicts as well.

**Comment by jon-wei** [comment]
Opened a new issue: https://github.com/apache/druid/pull/10015
